projects/ADuCM3029_demo_cn0415/src/cn0415.c

```diff
@@ -0,0 +1,2189 @@
+{
+	uint16_t freq;
+
+	freq = atoi((char *)args);
+
+	/* Update sampling frequency */
+	pid_set_sample_frequency(dev->controller, freq);
+
+	/* Save PID constants in non-volatile memory */
+	dev->param_array[F_LOOP] = dev->controller->f_loop;
+	dev->param_array[PID_KI_INT] = dev->controller->ki;
+	dev->param_array[PID_KI_DEC] =
+		(uint32_t)(dev->controller->ki * 10000) % 10000;
+	dev->param_array[PID_KD_INT] = dev->controller->kd;
+	dev->param_array[PID_KD_DEC] =
+		(uint32_t)(dev->controller->kd * 10000) % 10000;
+	return flash_write(dev->param_array, FLASH_PARAM_SIZE);
+}
+
+/***************************************************************************//**
+ * @brief Set set-point for the PID controller.
+ *
+ * @param dev  - The device structure.
+ * @param args - The new set-point.
+ *
+ * @return CN415_SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t cn0415_pid_set_setpoint(struct cn0415_dev *dev, uint8_t *args)
+{
+	float set_point;
+
+	set_point = atof((char *)args);
+
+	/* Set current set-point */
+	pid_set_setpoint(dev->controller, set_point);
+
+	/* Save set-point in non-volatile memory */
+	dev->param_array[HOLD_VAL_INT] = dev->controller->setpoint;
+	dev->param_array[HOLD_VAL_DEC] =
+		(uint32_t)(dev->controller->setpoint * 10000) % 10000;
+	return flash_write(dev->param_array, FLASH_PARAM_SIZE);
+}
+
+/***************************************************************************//**
+ * @brief Set duty cycle for the PWM.
+ *
+ * @param dev 	   - The device structure.
+ * @param new_duty - The new duty cycle.
+ *
+ * @return CN415_SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t cn0415_pwm_set_duty(struct cn0415_dev *dev, uint8_t* new_duty)
+{
+	uint16_t iduty;
+
+	/* Convert new duty cycle value from ASCII to integer representation */
+	iduty = (uint16_t)atoi((char*)new_duty);
+
+	/* Change PWM duty cycle */
+	return pwm_set_duty_cycle(dev->aducm3029_pwm_desc, iduty);
+}
+
+/***************************************************************************//**
+ * @brief Set frequency for the PWM.
+ *
+ * @param dev  - The device structure.
+ * @param freq - The new frequency.
+ *
+ * @return CN415_SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t cn0415_pwm_set_freq(struct cn0415_dev *dev, uint8_t* freq)
+{
+	int32_t ret;
+	uint32_t ifreq;
+
+	/* Convert new frequency value from ASCII to integer representation */
+	ifreq = (uint32_t)atoi((char*)freq);
+
+	/* Change PWM frequency */
+	ret = pwm_set_frequency(dev->aducm3029_pwm_desc, ifreq);
+	if(ret != 0)
+		return ret;
+
+	/* Save frequency in flash */
+	dev->param_array[F_PWM] = ifreq;
+	return flash_write(dev->param_array, FLASH_PARAM_SIZE);
```

Comment:
nitpick: I think the convention is to leave a white line before return:
```
	dev->param_array[F_PWM] = ifreq;

	return flash_write(dev->param_array, FLASH_PARAM_SIZE);
```

Check this everywhere in your code